### PR TITLE
Another batch of improvements :)

### PR DESCRIPTION
**Pull request by guizmaii**
And @sderosiaux worked on simplifying your `GameResult.collectRegion` function.

He'll push you a PR too ;)

(as usual, can you please squash my commits ? thanks 🙂)

### CODE REVIEW

File: src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala [hunk 1]
```diff
@@ -67,14 +67,17 @@ class GoBoard(val row: Int, val col: Int) {
...
 
   private def removeString(goString: GoString): Unit =
-    for (point <- goString.stones) {
-      // Removing a string can create liberties for other strings.
-      for (neighbor <- neighborMap((point._1, point._1)) if grid.get(neighbor.toCoords).isDefined) {
-        val neighborString = grid.get(neighbor.toCoords)
-        if (neighborString.get.equals(goString))
-          this.replaceString(neighborString.get.withLiberty(Point(point._1, point._2)))
+    goString.stones.foreach { point =>
+      neighborMap((point._1, point._1)).foreach { neighbor =>
```
Comment (guizmaii) [inline src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala]: is it normal that we pass `(point._1, point._1)` to `neighborMap` here ?

File: src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala [hunk 2]
```diff
@@ -53,64 +96,62 @@ object GameResult {
...
             }
-          }
         }
       }
     }
-    new Territory(statusMap)
+    statusMap.toMap
   }
 
+  private[this] final val deltas = List((-1, 0), (1, 0), (0, -1), (0, 1))
```
Comment (guizmaii) [inline src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala]: Don't loose time to review the code following this line. 

@sderosiaux has a PR for you simplifying the `collectRegion` function.

**Comment by sderosiaux** [comment]
Thanks @guizmaii. Here it is https://github.com/maxpumperla/ScalphaGoZero/pull/10.

File: src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala (hunk 1 above)
Comment (maxpumperla) [reply src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala]: omg, no! good catch.

File: src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala [hunk 3]
```diff
@@ -34,15 +57,35 @@ object GameResult {
...
+  def computeGameResult(goBoard: GoBoard): GameResult = {
+    val territoryMap = evaluateTerritory(goBoard)
+
+    var numBlackStones = 0
+    var numWhiteStones = 0
+    var numBlackTerritory = 0
+    var numWhiteTerritory = 0
+    var numDame = 0
+
+    for ((_, status) <- territoryMap) {
```
Comment (sderosiaux) [inline src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala]: FYI @guizmaii the best way here would be to import cats: `territoryMap.values.foldMap(t => Map(t -> 1))` (it uses the `Int` `Monoid` and return a `Map[Type, Int]`), and probably use cats in more places.

Comment (maxpumperla) [reply src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala]: @guizmaii @sderosiaux let's leave cats out, please. You might see this differently, but it's often the beginning of the end of a project's readability :)

**Comment by maxpumperla** [comment]
@guizmaii assuming you continue your refactoring odyssee, would you mind branching from current master? It becomes increasingly difficult to understand what your latest PR added. :/

File: src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala (hunk 3 above)
Comment (sderosiaux) [reply src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala]: Sure @maxpumperla, no worries.

Comment (guizmaii) [reply src/main/scala/org/deeplearning4j/scalphagozero/scoring/GameResult.scala]: @sderosiaux thanks ! I didn't know that :) but as requested. I'll not add Cats ;)

File: src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala (hunk 1 above)
Comment (guizmaii) [reply src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala]: Would like me to fix this or you'll fix it ?

**Comment by guizmaii** [comment]
@maxpumperla It's my last PR. I'll not have any more time to spend on this project until a long time. I have a lot more work to do on other projects. 

I hope that these PRs were useful to you. If you have any question on Scala or FP you can always contact me on Twitter. 

It was fun to work on your project. I don't know Go and discovered some rules by reading the code. 🙂

Again, I don't want to be cocky but please try to avoid mutability as much as possible. It'll ease the understanding of your work to others and it'll help you to not implement bugs. Scala is a great language for that.

File: src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala [hunk 4]
```diff
@@ -85,31 +88,26 @@ class GoBoard(val row: Int, val col: Int) {
...
+        case _                                                        => ()
+      }
     }
-    var allNeighborsInDanger = true
-    for (neighbor: GoString <- friendlyStrings)
-      if (neighbor.numLiberties != 1) allNeighborsInDanger = false
-    if (allNeighborsInDanger) return true
-    false
+
+    friendlyStrings.forall(_.numLiberties == 1)
```
Comment (guizmaii) [inline src/main/scala/org/deeplearning4j/scalphagozero/board/GoBoard.scala]: could be improved:

```scala
! friendlyStrings.exists(_.numLiberties != 1)
```

**Pull request merged**